### PR TITLE
[cxx-interop] Support forward declared records inside other records.

### PR DESCRIPTION
**Pull request by zoecarver**
The pattern:
```
    struct X { friend struct Y; }; struct Y {};
```
is fairly common. But before this commit, we would crash while attempting to add "Y" as a child of "X". This commit simply checks if the child record is a declaration or definition. If the former, it bails (and the "child" record will be imported where it's defined).

### CODE REVIEW

**Review by varungandhi-apple — APPROVED** [review]

**Comment by zoecarver** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ImportDecl.cpp
Comment (hlopko) [inline lib/ClangImporter/ImportDecl.cpp:3372]: Could you add a test for forward declaration without definition?

Are you only interested in skipping friend declarations? If so we could make the check more specific.

Will your change do the right thing (which I guess is importing the definition) for this snippet?

```cc
struct Foo {
  struct Bar;
};

struct Foo::Bar {
  int a;
};
```

Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:3372]: I'll keep the current test and add this one. I think it's beneficial to have both because the friend declaration isn't scoped.

Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:3372]: I tried adding that test but it, unfortunately, requires #34869. Once that lands, I'll update this patch with your suggested test.

**Comment by zoecarver** [comment]
After rebasing this is now a test-only change.

**Comment by zoecarver** [comment]
This is now crashing in a different place. #35085 should fix the new crash, so I'm just going to wait for that to land.

File: test/Interop/Cxx/class/Inputs/nested-records.h
Comment (zoecarver) [inline test/Interop/Cxx/class/Inputs/nested-records.h:77]: Note: if `ForwardDeclaredFriend` is outside of the namespace, there's no crash.

File: test/Interop/Cxx/class/Inputs/nested-records.h
Comment (gribozavr) [inline test/Interop/Cxx/class/Inputs/nested-records.h:60]:
```suggestion
  friend void takesFriend(ForwardDeclaredFriend f);
```
Not needed in C++.

File: test/Interop/Cxx/class/Inputs/nested-records.h
Comment (gribozavr) [inline test/Interop/Cxx/class/Inputs/nested-records.h:75]:
```suggestion
inline void takesHasNestedForwardDeclaration(HasNestedForwardDeclaration) { }
```

File: test/Interop/Cxx/class/Inputs/nested-records.h
Comment (gribozavr) [inline test/Interop/Cxx/class/Inputs/nested-records.h:56]:
```suggestion
namespace NestedDeclIsAFirstForwardDeclaration {
```

File: test/Interop/Cxx/class/nested-records-module-interface.swift
Comment (gribozavr) [inline test/Interop/Cxx/class/nested-records-module-interface.swift:75]: Should `HasNestedForwardDeclaration` also appear here?

File: test/Interop/Cxx/class/Inputs/nested-records.h
Comment (zoecarver) [reply test/Interop/Cxx/class/Inputs/nested-records.h:60]: I know. This was part of the reproducer, IIRC it didn't crash without it. I can't verify this now that it no longer crashes, but I think we should keep it.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test macOS.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Comment by zoecarver** [comment]
Another Java error.

@swift-ci please test Windows.

**Comment by zoecarver** [comment]
The windows test failed because of an unrelated error. I'm going to go ahead and land this because it's just a NFC and it passed on the windows CI (even though another, unrelated test failed).

**Pull request merged**
